### PR TITLE
feat(op): pass unverifiable ENRs to rlpx

### PR DESCRIPTION
**Pull request by emhane**
Uses new `discv5::Event::UnverifiableEnr` to bridge CL and EL discv5 implementation

### CODE REVIEW

File: crates/net/discv5/src/lib.rs [hunk 1]
```diff
@@ -299,19 +312,20 @@ impl Discv5 {
...
     pub fn try_into_reachable(
         &self,
         enr: &discv5::Enr,
         socket: SocketAddr,
     ) -> Result<NodeRecord, Error> {
         let id = enr_to_discv4_id(enr).ok_or(Error::IncompatibleKeyType)?;
 
-        let udp_socket = self.ip_mode().get_contactable_addr(enr).unwrap_or(socket);
+        // todo: when impl ipv6 support in rlpx, favour advertised socket if ipv6 wrt to local ip
+        // mode
```
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:328]: is there anything we should be tracking for this?

unclear what favoring means here in this context, would lean towards removing this.

File: crates/net/discv5/src/lib.rs [hunk 2]
```diff
@@ -242,13 +242,24 @@ impl Discv5 {
...
                 self.on_discovered_peer(&enr, remote_socket)
             }
-            _ => None,
+            discv5::Event::UnverifiableEnr {
+                enr,
+                socket,
+                node_id: _,
+            } => {
+                // covers `reth_discv4::DiscoveryUpdate` equivalents `DiscoveryUpdate::Added(_)`
+                // and `DiscoveryUpdate::DiscoveredAtCapacity(_)
```
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:255]: unclear what the references to the types mean here

please clarify what covers + equivalents means here

**Review by mattsse — CHANGES_REQUESTED** [review]
jus some doc questions

File: crates/net/discv5/src/lib.rs (hunk 1 above)
Comment (emhane) [reply crates/net/discv5/src/lib.rs:328]: was thinking about dual stack here, think eventually, rlpx will have to support dual stack too: ipv6 and ipv4 sockets in parallel. consider it removed.

File: crates/net/discv5/src/lib.rs [hunk 3]
```diff
@@ -236,19 +236,32 @@ impl Discv5 {
...
+                socket,
+                node_id: _,
+            } => {
+                // this branch is semantically similar to branches of 
+                // `reth_discv4::DiscoveryUpdate`: `DiscoveryUpdate::Added(_)` and
+                // `DiscoveryUpdate::DiscoveredAtCapacity(_)
+
+                // peer has been discovered as part of query, or, by an outgoing session (but peer 
+                // is behind NAT and responds from a different socket)
+
```
Comment (joshieDo) [inline crates/net/discv5/src/lib.rs:261]: can we add a trace/debug here?

**Review by mattsse — APPROVED** [review]

File: crates/net/discv5/src/lib.rs [hunk 4]
```diff
@@ -299,20 +314,18 @@ impl Discv5 {
...
     /// Tries to convert an [`Enr`](discv5::Enr) into the backwards compatible type [`NodeRecord`],
-    /// w.r.t. local [`IpMode`]. Tries the socket from which the ENR was sent, if socket is missing
-    /// from ENR.
+    /// w.r.t. local [`IpMode`]. Uses source socket as udp socket.
     ///
-    ///  Note: [`discv5::Discv5`] won't initiate a session with any peer with a malformed node
-    /// record, that advertises a reserved IP address on a WAN network.
+    ///  Note: [`discv5::Discv5`] won't initiate a session with any peer with an unverifiable node
+    /// record, for example one that advertises a reserved LAN IP address on a WAN network. This is
+    /// in order to prevent DoS attacks, where some malicious peers may advertise a victim's socket.
```
Comment (joshieDo) [inline crates/net/discv5/src/lib.rs:321]: with this we do though. could we just specify that in such cases we try into `socket` instead of the enr record

**Review by joshieDo — APPROVED** [review]
just couple nits, but lgtm

did you notice an improvement ?

File: crates/net/discv5/src/lib.rs (hunk 4 above)
Comment (emhane) [reply crates/net/discv5/src/lib.rs:321]: hm that's not a discv5 session that would result, that would be an rlpx session

**Comment by emhane** [comment]
> just couple nits, but lgtm
> 
> did you notice an improvement ?

have ran it. doesn't interfere with regular flow, and is purely an enhancement. so will not cause worse behaviour.

**Pull request merged**